ably/realtime_conn.go

```diff
@@ -664,6 +664,22 @@ func (c *Connection) setSerial(serial int64) {
 	c.serial = serial
 }
 
+func (c *Connection) resendPending() {
+	lg := c.logger().sugar()
+	c.mtx.Lock()
+	cx := make([]msgCh, len(c.pending.queue))
+	copy(cx, c.pending.queue)
+	c.pending.queue = []msgCh{}
+	c.mtx.Unlock()
+	lg.Debugf("resending %d messages waiting for ACK/NACK", len(cx))
+	for _, v := range cx {
+		err := c.send(v.msg, v.ch)
+		if err != nil {
+			lg.Errorf("failed to re send message %v with error:%v", v.msg, err)
```

Comment:
Should the message not be re-queued?

Reply:
`send` [requeues the message and returns nil](https://github.com/ably/ably-go/blob/2c4d62974472f5f8178dc20ba0e90935aa13ec2b/ably/realtime_conn.go#L584-L585), following [RTL6c](https://docs.ably.io/client-lib-development-guide/features/#RTL6c).

_But_ that only works if the connection is already disconnected. If it gets disconnected as a result of the send operation failing, the message won't be requeued, nor enqueued in the first place.

It seems to me that `send` shouldn't return an error at all. The error feedback will be sent to the channel it takes, eventually. But its job can't fail —it either succeeds sending the message, succeeds enqueueing it or succeeds sending an error through the channel (that is handled elsewhere).

This problem predates this PR, so maybe it can be fixed in another PR?

Reply:
Ok, but I asked the same question [here](https://github.com/ably/ably-go/pull/208#discussion_r519712866) and you said this PR covers fixing it, so it's fine if this PR _doesn't_ fix it, but we really should fix it 😉

Reply:
@lmars  I think you bring up a good point here. This is a direct send .

@tcard I think we  might need to address it here. The logic was supposed to be same as `RealtimeChannel.send` which takes care of queuing and address @lmars point. We just don't have reference to the original Channel that sent the message at this point , need to figure out out how to pass that reference.

Reply:
Mmh! Thinking about it again, I think @tcard is right, I misunderstood the issue. 

@lmars 
>Should the message not be re-queued?
It depends, you should know that, by the time we resending stuff, we are in a new transport meaning we have acrive connection. So the message might be probably either failed  to `queue` or `sent`

But, I also think `send` should not return an error at all. Any error should be  sent to the `listen` channel

Reply:
Ok, from a high level we just shouldn't be dropping messages that fail to send, they should be re-queued and resent.

Reply:
@gernest have you been able to make any progress on this?

Reply:
LGTM thanks 👍